src/main/scala/encry/crypto/equihash/Equihash.scala

```diff
@@ -272,56 +205,53 @@ object Equihash {
     */
   @SuppressWarnings(Array("NullParameter"))
   def validateSolution(n: Char, k: Char, personal: Array[Byte], header: Array[Byte], solutionIndices: IndexedSeq[Int]): Boolean = {
+
     assert(n > 1)
     assert(k >= 3)
     assert(n % 8 == 0)
     assert(n % (k + 1) == 0)
 
-    val solutionLen = Math.pow(2, k).toInt
+    val solutionLen: Int = Math.pow(2, k).toInt
     assert(solutionIndices.size == solutionLen)
 
-    // Check for duplicate indices.
-    if (solutionIndices.toSet.size != solutionIndices.size) {
-      false
-    } else {
-      // Generate hash words.
-      val bytesPerWord = n / 8
-      val wordsPerHash = 512 / n
-      val outlen = wordsPerHash * bytesPerWord
+    if (solutionIndices.toSet.size != solutionIndices.size) false
+    else {
+      val bytesPerWord: Int = n / 8
+      val wordsPerHash: Int = 512 / n
+      val outlen: Int = wordsPerHash * bytesPerWord
 
-      val digest = new Blake2bDigest(null, outlen, null, personal)
+      val digest: Blake2bDigest = new Blake2bDigest(null, outlen, null, personal)
       digest.update(header, 0, header.length)
 
-      // Check pair-wise ordening of indices.
-      for (s <- 0 until k) {
-        val d = 1 << s
-        for (i <- 0 until solutionLen by 2 * d) {
-          if (solutionIndices(i) >= solutionIndices(i + d)) {
-            return false
-          }
+      val pairWiseCheck: Boolean = (0 until k).forall(s => {
+        val d: Int = 1 << s
+        (0 until solutionLen by 2 * d).forall(i =>
+          if (solutionIndices(i) >= solutionIndices(i + d)) false
+          else true
+        )
+      })
+
+      val words: ArrayBuffer[BigInteger] =
+        (0 until solutionLen).foldLeft(ArrayBuffer.empty[BigInteger]) {
+          case (buffer, i) => buffer += generateWord(n, digest, solutionIndices(i))
         }
-      }
-
-      val words = ArrayBuffer.empty[BigInteger]
-      for (i <- 0 until solutionLen) {
-        words += generateWord(n, digest, solutionIndices(i))
-      }
 
-      // Check XOR conditions.
-      val bitsPerStage = n / (k + 1)
-      for (s <- 0 until k) {
-        val d = 1 << s
-        for (i <- 0 until solutionLen by 2 * d) {
-          val w = words(i).xor(words(i + d))
-          if (w.shiftRight(n - (s + 1) * bitsPerStage) != BigInteger.ZERO) {
-            return false
-          }
-          words(i) = w
-        }
+      val xorConditionsCheck: Boolean = {
```

Comment:
A lot of using cycles and mutable variables in file. 
Logic is not transparent at all. In future it should be refactored. I think makes sense to create task for this. 
Now approve.